grails-app/services/org/pih/warehouse/inventory/ProductAvailabilityService.groovy

```diff
@@ -562,4 +564,79 @@ class ProductAvailabilityService {
         log.info "Updated ${results} product availability records for product ${product?.productCode}"
     }
 
+
+    /**
+     *
+     * @param commandInstance
+     * @return
+     */
+    List searchProducts(InventoryCommand command) {
+        def categories = inventoryService.getExplodedCategories([command.category])
+        List searchTerms = (command?.searchTerms ? Arrays.asList(command?.searchTerms?.split(" ")) : null)
+
+        // Only search if there are search terms otherwise the list of product IDs includes all products
+        def innerProductIds = !searchTerms ? [] : Product.createCriteria().list {
+            eq("active", true)
+            projections {
+                distinct 'id'
+            }
+            and {
+                searchTerms.each { searchTerm ->
+                    or {
+                        ilike("name", "%" + searchTerm + "%")
+                        inventoryItems {
+                            ilike("lotNumber", "%" + searchTerm + "%")
+                        }
+                    }
+                }
+            }
+        }
+
+        def products = Product.createCriteria().list {
+            eq("active", true)
+            and {
+                if (categories) {
+                    'in'("category", categories)
+                }
+                if (command.tags) {
+                    tags {
+                        'in'("id", command.tags*.id)
+                    }
+                }
+                if (command.catalogs) {
+                    productCatalogItems {
+                        productCatalog {
+                            'in'("id", command.catalogs*.id)
+                        }
+                    }
+                }
+                // This is pretty inefficient if the previous query does not narrow the results
+                if (innerProductIds) {
+                    'in'("id", innerProductIds)
+                }
+            }
+        }
+
+        // pa.quantityOnHand > 0 check should be done with check for pa.product.productType.supportedActivities = SEARCHABLE
+
+        def productsWithQoH = ProductAvailability.executeQuery("""
```

Comment:
Is ProductSummary the same case as ProductAvailability?

Reply:
@jmiranda I tried with the product availability left outer join to the product query. It works, but I had to add product.productAvailability association (via hasMany) for it to work (because gorm/hql requires association for joins). However, the `No value specified for parameter 13` is still present when I try to look into supported activities.